### PR TITLE
[MODGOBI-195] - error handling improvement

### PR DESCRIPTION
**Pull request by imerabishvili**
## Purpose
https://github.com/folio-org/mod-gobi/tree/MODGOBI-195

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java [hunk 1]
```diff
@@ -41,18 +21,30 @@
...
+import static org.folio.rest.ResourcePaths.CONFIGURATION_ENDPOINT;
+import static org.folio.rest.ResourcePaths.ORDERS_BY_ID_ENDPOINT;
+import static org.folio.rest.ResourcePaths.ORDERS_ENDPOINT;
+import static org.folio.rest.jaxrs.resource.Gobi.PostGobiOrdersResponse.respond400WithApplicationXml;
+import static org.folio.rest.jaxrs.resource.Gobi.PostGobiOrdersResponse.respond401WithTextPlain;
+import static org.folio.rest.jaxrs.resource.Gobi.PostGobiOrdersResponse.respond500WithApplicationXml;
 
 public class PostGobiOrdersHelper {
-  private static final Logger logger = LogManager.getLogger(PostGobiOrdersHelper.class);
+  private static final Logger LOGGER = LogManager.getLogger(PostGobiOrdersHelper.class);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:47]: In all othe places logger defined with lower case? In this case I think it does not makes sense convention to full module

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java [hunk 2]
```diff
@@ -121,71 +112,82 @@ public CompletableFuture<Document> parseXML(String entity) {
...
+          String config = jo.getJsonArray(CONFIGS).getJsonObject(0).getString("value");
+          LOGGER.info("lookupOrderMappings:: Use custom mapping config: \n {}", config);
+          return tryExtractOrderMappings(orderType, jo);
+        } else {
+          String config = JsonObject.mapFrom(MappingHelper.getDefaultMappingByOrderType(orderType)).encode();
+          LOGGER.info("lookupOrderMappings:: Use default mapping config: \n {}", config);
+          return getDefaultMappingsFromCache(orderType);
+        }
+      })
+      .onFailure(t -> LOGGER.error("Exception looking up custom order mappings for tenant '{}'", restClient.getTenantId()))
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:140]: lets add error message to this log as well

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java [hunk 3]
```diff
@@ -248,75 +250,78 @@ private CompletableFuture<Boolean> checkExistingOrder(CompositePurchaseOrder com
...
+    }
+
+    return respond500WithApplicationXml(message);
   }
+
+  private static BinaryOutStream writeGobiResponse(String errorCode, String message) {
+    GobiResponse response = new GobiResponse();
+    response.setError(new ResponseError());
+    response.getError().setCode(errorCode);
+    response.getError().setMessage(message == null ? null : message.substring(0, Math.min(message.length(), 500)));
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:323]: thinking that we don't need to reduce lenght of message here

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java [hunk 4]
```diff
@@ -248,75 +250,78 @@ private CompletableFuture<Boolean> checkExistingOrder(CompositePurchaseOrder com
...
-    asyncResultHandler.handle(Future.succeededFuture(result));
-    return null;
+    if (t instanceof GobiPurchaseOrderParserException) {
+      return respond400WithApplicationXml(writeGobiResponse(CODE_INVALID_XML, message));
+    }
+
+    if (t instanceof GobiException) {
+      String errorCode = ((GobiException) t).getErrorCode().getCode();
+      return respond500WithApplicationXml(writeGobiResponse(errorCode, message));
+    }
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:314]: You can rewrite with java 17
```
if (t instanceof GobiException gobiException) {
    String errorCode = gobiException.getErrorCode().getCode();
    return respond500WithApplicationXml(writeGobiResponse(errorCode, message));
}
```

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java (hunk 3 above)
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:323]:
```
response.getError().setMessage(StringUtils.substring(message, 0, MAX_MESSAGE_LENGTH));
```
This code also handle null, but interesting why this limit of 500 symbols needed, for me better to remove it

**Comment by azizbekxm** [comment]
@imerabishvili Thanks to great changes, especially additional refactoring. 
One thing, please update your code style settings especially import styles
https://easyupload.io/9fjk8t that my full java style (it may not be necessary to use all of them settings ) 
To be compatible with across all module import use these settings import file to your code style 
image
image
image
image
image
image

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java [hunk 5]
```diff
@@ -248,75 +250,78 @@ private CompletableFuture<Boolean> checkExistingOrder(CompositePurchaseOrder com
...
-        break;
-      case 500:
-        result = respond500WithTextPlain(message);
-        break;
-      case 401:
-        result = respond401WithTextPlain(message);
-        break;
-      default:
-        result = respond500WithTextPlain(message);
+      switch (((HttpException) t).getCode()) {
```
Comment (azizbekxm) [inline src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:299]: these style can be improved by using java 17 style

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java (hunk 4 above)
Comment (imerabishvili) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:314]: As I checked patterns in switch are supported from v19, can you double check?

Comment (azizbekxm) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:314]: updating java 17 pr for mod-orders https://github.com/folio-org/mod-orders/pull/760/files for reference

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java (hunk 3 above)
Comment (imerabishvili) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:323]: It's not the stacktrace to care about response size so removing it...

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java (hunk 5 above)
Comment (azizbekxm) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:299]: check all import settings, especially their order

Comment (imerabishvili) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:299]: Thank you. The only difference now is that currentl java.* is followed by javax.* but in your screenshots they have a reverse order. Does it make sense to reformat those?

Comment (azizbekxm) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:309]: javax.
java.
is correct one.

Comment (azizbekxm) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:309]: I am not fan of these style to be honest. it is decided by folio based on Eclipse idea a while ago. I use this import style to be more compatible with all modules

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java (hunk 4 above)
Comment (imerabishvili) [reply src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:324]: We're already using java 17. The expression:
```
return switch (someInt) {
  case 1 -> 1;
  default -> 2;
};
```
is good however following is not:
```
return switch (someObject) {
  case Integer i -> 1;
  case Double d -> 2;
  default -> 3;
};
```

File: src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java [hunk 6]
```diff
@@ -281,42 +293,45 @@ private CompletableFuture<CompositePurchaseOrder> getExistingOrderById(Composite
       });
 
   }
+
   public Void handleError(Throwable throwable) {
-    final javax.ws.rs.core.Response result;
+    Throwable cause = throwable.getCause();
+    logger.error("Exception placing order", cause);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java:299]: Better to log thowable here to have all info with full stack trace

**Review by SerhiiNosko — APPROVED** [review]

**Pull request merged**